File: Astro-TFG/public/scripts/2-procesamiento-datos/prueba-1/version-1/python/main.py
import math
import time
import random
import tracemalloc
import gc
import js  # type: ignore
from pyscript import display


def create_data_structure(my_list, size):
    for _ in range(size):
        my_list.append(random.randint(0, 1000))


def transform_data_structure(my_list):
    aux_list = my_list.copy()
    for i in range(len(aux_list)):
        aux_list[i] = (aux_list[i] ** 2 + math.log(aux_list[i] + 1)
                       ) / math.sqrt(aux_list[i] + 1)


def sort_data_structure(my_list):
    aux_list = my_list.copy()
    aux_list.sort()


def search_in_data_structure(my_list, value):
    for item in my_list:
        if item == value:
            return True
    return False


def filter_data_structure(my_list, threshold):
    return [x for x in my_list if x > threshold]


def delete_from_data_structure(my_list, value):
    i = 0
    while i < len(my_list):
        if my_list[i] == value:
            my_list.pop(i)
        else:
            i += 1


def get_memory_usage():
    return tracemalloc.get_traced_memory()[1] / (1024 * 1024)


def do_operations(size):
    gc.collect()
    tracemalloc.start()
    my_list = []
    metrics = {}

    start_total = time.perf_counter()
    start_op = time.perf_counter()

    create_data_structure(my_list, size)

    tracemalloc.stop()

    metrics['create'] = {
        'time': (time.perf_counter() - start_op) * 1000,
        'memory': get_memory_usage()
    }

    operations = [
        ('transform', transform_data_structure, (my_list,)),
        ('sort', sort_data_structure, (my_list,)),
        ('search', search_in_data_structure, (my_list, random.choice(my_list))),
        ('filter', filter_data_structure, (my_list, 500)),
        ('delete', delete_from_data_structure, (my_list, random.choice(my_list)))
    ]

    for op_name, op_func, args in operations:
        gc.collect()
        tracemalloc.start()
        start_op = time.perf_counter()

        op_func(*args)

        metrics[op_name] = {
            'time': float((time.perf_counter() - start_op) * 1000),
            'memory': get_memory_usage()
        }

        tracemalloc.stop()

    metrics['output'] = {
        'total_time': float((time.perf_counter() - start_total) * 1000),
        'memory_peak': max([op['memory'] for op in metrics.values() if 'memory' in op])
    }
    display_results(metrics)


def display_results(results):
    for op, data in results.items():
        if op != 'output':
            display(
                f"{op.upper()} - Time: {data['time']:.2f} ms | RAM: {data['memory']:.2f} MB",
                target="pyscript-output"
            )
    display(
        f"Total ET: {results['output']['total_time']:.2f} ms", target="pyscript-output"
    )
    display(
        f"RAM Peak: {results['output']['memory_peak']:.2f} MB", target="pyscript-output"
    )
    js.hideExecutionLoader()


def run_py_benchmark(event):
    js.clearCell('pyscript-output')
    do_operations(10_000_000)


js.window.run_py_benchmark = run_py_benchmark

File: Astro-TFG/public/scripts/2-procesamiento-datos/prueba-1/version-1/javascript/main.js
function createDataStructure(my_list, size) {
    for (let i = 0; i < size; i++) {
        my_list.push(Math.floor(Math.random() * 1001));
    }
}

function transformDataStructure(my_list) {
    const aux_list = [...my_list];
    for (let i = 0; i < aux_list.length; i++) {
        const num = aux_list[i];
        aux_list[i] = (Math.pow(num, 2) + Math.log(num + 1)) / Math.sqrt(num + 1);
    }
}

function sortDataStructure(my_list) {
    const aux_list = [...my_list];
    aux_list.sort((a, b) => a - b);
}

function searchInDataStructure(my_list, value) {
    for (let i = 0; i < my_list.length; i++) {
        if (my_list[i] === value) {
            return true;
        }
    }
    return false;
}

function filterDataStructure(my_list, threshold) {
    const filtered = [];
    for (let i = 0; i < my_list.length; i++) {
        if (my_list[i] > threshold) {
            filtered.push(my_list[i]);
        }
    }
    return filtered;
}

function deleteFromDataStructure(my_list, value) {
    let i = 0;
    while (i < my_list.length) {
        if (my_list[i] === value) {
            my_list.splice(i, 1);
        } else {
            i++;
        }
    }
}

let max_memory = 0;

function doOperations(size) {
    const my_list = [];
    const metrics = {};
    const start_total = performance.now();

    const measureMemory = () => {
        const current = performance.memory.usedJSHeapSize / (1024 * 1024);
        max_memory = Math.max(max_memory, current);
        return current;
    };

    let start_op = performance.now();
    createDataStructure(my_list, size);
    metrics['create'] = {
        time: performance.now() - start_op,
        memory: measureMemory()
    };

    const operations = [
        ['transform', transformDataStructure, [my_list]],
        ['sort', sortDataStructure, [my_list]],
        ['search', searchInDataStructure, [my_list, my_list[Math.floor(Math.random() * my_list.length)]]],
        ['filter', filterDataStructure, [my_list, 500]],
        ['delete', deleteFromDataStructure, [my_list, my_list[Math.floor(Math.random() * my_list.length)]]]
    ];

    for (const [op_name, op_func, args] of operations) {
        start_op = performance.now();
        if (window.gc) window.gc();

        const mem_before = measureMemory();
        op_func(...args);
        const mem_after = measureMemory();

        metrics[op_name] = {
            time: performance.now() - start_op,
            memory: Math.abs(mem_after - mem_before)
        };
    }

    metrics['output'] = {
        total_time: performance.now() - start_total,
        memory_peak: max_memory
    };

    const outputElement = document.getElementById("javascript-output");
    if (!outputElement) return;
    outputElement.innerHTML = "";

    for (const [op, data] of Object.entries(metrics)) {
        if (op === 'output') continue;
        const line = `${op.toUpperCase()} - Time av. : ${data.time.toFixed(2)} ms | RAM: ${data.memory.toFixed(2)} MB`;
        outputElement.innerHTML += line + "<br>";
    }

    outputElement.innerHTML += "<br>";
    const tot = metrics['output'];
    const totalLine = `TOTAL - Time: ${tot.total_time.toFixed(2)} ms | RAM Peak: ${tot.memory_peak.toFixed(2)} MB`;
    outputElement.innerHTML += totalLine;
}

window.runJsBenchmark = function () {
    clearCell("javascript-output");
    window.showExecutionLoader();

    requestAnimationFrame(() => {
        setTimeout(() => {
            doOperations(10_000_000);
            window.hideExecutionLoader();
        }, 0);
    });
};
